ratbag_emu/device.py

```diff
@@ -0,0 +1,272 @@
+        for keys in [a.keys for a in val]:
+            for el in keys:
+                assert el not in seen
+                seen.append(el)
+
+        self._actuators = val
+
+    def destroy(self):
+        for endpoint in self.endpoints:
+            endpoint.destroy()
+
+    def transform_action(self, data: Dict[str, Any]):
+        '''
+        Transforms high-level action according to the actuators
+
+        A high-level action will have the x, y values in mm. This values will
+        be converted to pixel by the device actuators (in this case, the
+        sensor/dpi actuator)
+
+        :param action:  high-level action
+        '''
+        hid_data: Dict[str, Any] = {}
+
+        for actuator in self.actuators:
+            hid_data.update(actuator.transform(data.copy()))
+
+        return hid_data
+
+    def send_hid_action(self, action: object):
+        '''
+        Sends a HID action
+
+        We assume there's only one endpoint for each type of action (mouse,
+        keyboard, button, etc.) so we send the action to all endpoints. The
+        endpoint will only act on the action if it supports it.
+
+        :param action:  HID action
+        '''
+        for endpoint in self.endpoints:
+            # FIXME: global data (0x11)?
+            endpoint.send(endpoint.create_report(action, 0x11))
+
+    def simulate_action(self, action: Dict[str, Any], type: int = None):
+        '''
+        Simulates action
+
+        Translates physical values according to the device properties and
+        converts action into HID reports.
+
+        :param action:  high-level action
+        :param type:    HID report type
+        '''
+
+        packets: Dict[int, EventData] = {}
+
+        report_count = int(round(ms2s(action['duration']) * self.report_rate))
+
+        if not report_count:
+            report_count = 1
+
+        # XY movement
+        if action['type'] == 'xy':
+            # We assume a linear motion
+            pixel_buffer = {}
+            step = {}
+
+            '''
+            Initialize pixel_buffer, real_pixel_buffer and step for X and Y
+
+            pixel_buffer holds the ammount of pixels left to send (kinda,
+            read below).
+
+            We actually have two variables for this, real_pixel_buffer and
+            pixel_buffer. pixel_buffer mimics the user movement and
+            real_pixel_buffer holds true number of pixels left to send.
+            When using high report rates (ex. 1000Hz) we usually don't
+            have a full pixel to send, that's why we need two variables. We
+            subtract the step to pixel_buffer at each iteration, when the
+            difference between pixel_buffer and real_pixel_buffer is equal
+            or higher than 1 pixel we then send a HID report to the device
+            with that difference (int! we send the int part of the
+            difference) and update real_pixel_buffer to include this
+            changes.
+            '''
```

Comment:
nitpick: using "pixel" is misleading. We are dealing with a physical mouse, so there is no notion of "pixel". "pixel" is a screen/display notion when here we are only dealing with physical properties.

The term `DPI` stands for `Dots Per Inch`, so we rather use `dots` in place of `pixels`